### PR TITLE
Rd 1485 update deployment apis

### PR DESCRIPTION
**Pull request by mabuaisha**
This PR created in order to update the following:
1. Add support to attach deployment(services)/environments to deployment object
2. Add support to include new extra fields as part of the deployment response related to environments
3. Add support to update deployments count for sub-deployments (environments && services) in API response

### CODE REVIEW

File: rest-service/manager_rest/resource_manager.py [hunk 1]
```diff
@@ -56,8 +56,11 @@
                                 extract_host_agent_plugins_from_plan)
 from manager_rest.rest.rest_utils import (
     RecursiveDeploymentDependencies,
-    update_inter_deployment_dependencies, verify_blueprint_uploaded_state,
-    compute_rule_from_scheduling_params)
+    RecursiveDeploymentLabelsDependencies,
+    update_inter_deployment_dependencies,
+    verify_blueprint_uploaded_state,
+    compute_rule_from_scheduling_params
```
Comment (geokala) [inline rest-service/manager_rest/resource_manager.py:62]: A trailing comma would be nice here, but not critical at all.

File: rest-service/manager_rest/resource_manager.py [hunk 2]
```diff
@@ -1381,6 +1383,105 @@ def _finalize_create_deployment(self, deployment: models.Deployment):
...
+        ).items
+        _existing_parents = [_parent[0] for _parent in result]
+        for parent in parents:
+            if parent not in _existing_parents:
+                missing_parents.append(parent)
+
+        if missing_parents:
+            raise manager_exceptions.DeploymentParentNotFound(
+                'Deployment {0}: is referencing deployments'
+                ' using label `csys-obj-parent` that are not existed, '
```
Comment (geokala) [inline rest-service/manager_rest/resource_manager.py:1411]: that are not existed -> that do not exist

File: rest-service/manager_rest/resource_manager.py [hunk 3]
```diff
@@ -1381,6 +1383,105 @@ def _finalize_create_deployment(self, deployment: models.Deployment):
...
+                'make sure that deployments {1} are existed before create '
```
Comment (geokala) [inline rest-service/manager_rest/resource_manager.py:1412]: make sure that deployments {1} are existing before create -> make sure that deployment(s) {1} exist before creating

File: rest-service/manager_rest/resource_manager.py [hunk 4]
```diff
@@ -1381,6 +1383,105 @@ def _finalize_create_deployment(self, deployment: models.Deployment):
...
+                'deployment'.format(deployment_id, ','.join(missing_parents))
+            )
+
+    def _place_deployment_label_dependency(self, source, target):
+        now = utils.get_formatted_timestamp()
```
Comment (geokala) [inline rest-service/manager_rest/resource_manager.py:1417]: I don't think you need to provide this- it already defaults to the current time, so probably best just to not provide it except when it's being overridden (which this isn't).

File: rest-service/manager_rest/rest/resources_v3_1/deployments.py [hunk 5]
```diff
@@ -111,7 +170,7 @@ def put(self, deployment_id, **kwargs):
             skip_plugins_validation=skip_plugins_validation,
             site=site,
             runtime_only_evaluation=request_dict.get(
-                'runtime_only_evaluation', False),
```
Comment (geokala) [inline rest-service/manager_rest/rest/resources_v3_1/deployments.py:114]: Why remove the trailing comma here?

File: rest-service/manager_rest/rest/resources_v3_1/deployments.py [hunk 6]
```diff
@@ -130,13 +189,6 @@ def put(self, deployment_id, **kwargs):
                 raise self._error_from_create(deployment.create_execution)
         return deployment, 201
 
-    def _error_from_create(self, execution):
```
Comment (geokala) [inline rest-service/manager_rest/rest/resources_v3_1/deployments.py:133]: Why was this moved nearer the start of the file?

File: rest-service/manager_rest/rest/rest_utils.py [hunk 7]
```diff
@@ -644,57 +644,64 @@ def find_recursive_deployments(self, source_id):
...
-    def increase_deployment_counts_in_graph(self, target, source):
+    def increase_deployment_counts_in_graph(self, target_id, source):
         """
         Increase the deployment counts for target deployment that is
         referenced directly by the `source` deployment and then make sure to
         propagate that increase to all deployment that are referenced
         directly and indirectly by `target` deployment
-        :param target: Target Deployment node that we need to attach source to
-        :rtype Deployment
+        :param target_id: Target Deployment ID that we need to attach
```
Comment (geokala) [inline rest-service/manager_rest/rest/rest_utils.py:653]: I think it's better to continue to  pass the target deployment in here rather than the ID itself- we recently changed to this approach in several places.

File: rest-service/manager_rest/rest/rest_utils.py [hunk 8]
```diff
@@ -644,57 +644,64 @@ def find_recursive_deployments(self, source_id):
...
+        source to
+        :rtype str
         :param source: Source deployment that reference target deployment
         :rtype Deployment
         """
-        total_services = source.sub_services_count or 0
```
Comment (geokala) [inline rest-service/manager_rest/rest/rest_utils.py:658]: This (and the following line's change) seems to just make some of the following code more complicated- is there any reason to take this approach?

File: rest-service/manager_rest/rest/rest_utils.py [hunk 9]
```diff
@@ -644,57 +644,64 @@ def find_recursive_deployments(self, source_id):
...
+                self.sm.update(target)
 
-    def decrease_deployment_counts_in_graph(self, source):
+    def decrease_deployment_counts_in_graph(self, target_id, source):
         """
         Decrease the counter for each deployment that is referenced
         directly and indirectly by `source` deployment and the counts that
         need to be update are possibly for `sub_environments_count`
         & `sub_services_count`
+        :param target_id: Target Deployment ID that we need to de-attach
```
Comment (geokala) [inline rest-service/manager_rest/rest/rest_utils.py:690]: I think it'd be better to go with the approach of passing the deployment here (as noted above).

File: rest-service/manager_rest/rest/rest_utils.py [hunk 10]
```diff
@@ -644,57 +644,64 @@ def find_recursive_deployments(self, source_id):
...
         :rtype Deployment
         """
-        target_group = self.find_recursive_deployments(source.id)
+        target_group = [target_id] + self.find_recursive_deployments(target_id)
         if target_group:
             total_services = source.sub_services_count or 0
             total_environments = source.sub_environments_count or 0
             if source.is_environment:
-                total_environments += 1
+                total_environments = total_environments + 1 if \
```
Comment (geokala) [inline rest-service/manager_rest/rest/rest_utils.py:701]: You don't need this extra logic I think, because you have the 'or 0' lines above.

Comment (geokala) [reply rest-service/manager_rest/rest/rest_utils.py:701]: (so you can just do total_environments += 1)

File: rest-service/manager_rest/rest/rest_utils.py [hunk 11]
```diff
@@ -644,57 +644,64 @@ def find_recursive_deployments(self, source_id):
...
+                    total_environments else 1
             else:
-                total_services += 1
```
Comment (geokala) [inline rest-service/manager_rest/rest/rest_utils.py:697]: Similar to the previous comment, I don't see why this change is needed?

File: rest-service/manager_rest/rest/rest_utils.py [hunk 12]
```diff
@@ -703,8 +710,12 @@ def decrease_deployment_counts_in_graph(self, source):
                 )
                 if target.sub_services_count:
                     target.sub_services_count -= total_services
+                    if not target.sub_services_count:
+                        target.sub_services_count = None
```
Comment (geokala) [inline rest-service/manager_rest/rest/rest_utils.py:714]: Out of interest, and bearing all the previous logic in mind: Why are we ever allowing these to be null values- why not instead require them to be integers, defaulting to 0- that seems like it would simplify a lot of this code.

File: rest-service/manager_rest/rest/rest_utils.py [hunk 13]
```diff
@@ -703,8 +710,12 @@ def decrease_deployment_counts_in_graph(self, source):
...
                 if target.sub_environments_count:
                     target.sub_environments_count -= total_environments
+                    if not target.sub_environments_count:
```
Comment (geokala) [inline rest-service/manager_rest/rest/rest_utils.py:717]: Same note as previous comment- this should maybe not be allowing null values.

File: rest-service/manager_rest/storage/resource_models.py [hunk 14]
```diff
@@ -617,6 +631,25 @@ def make_create_environment_execution(self, **params):
...
+        return ''
+
+    @property
+    def finished_operations(self):
+        if not self.latest_execution:
+            return None
+        return self.latest_execution.finished_operations
+
+    @property
+    def total_operations(self):
```
Comment (geokala) [inline rest-service/manager_rest/storage/resource_models.py:648]: These two _operations properties feel somewhat misleading as they only give you finished/total operations for the most recent execution. I'm unsure if this is a big issue, but it doesn't feel quite right.

File: rest-service/manager_rest/test/endpoints/test_deployment_labels_dependencies.py [hunk 15]
```diff
@@ -0,0 +1,341 @@
...
+        deployment_2 = self.client.deployments.get('parent_2')
+        self.assertEqual(deployment_1.sub_services_count, 1)
+        self.assertEqual(deployment_1.sub_environments_count, 0)
+        self.assertEqual(deployment_2.sub_services_count, 1)
+        self.assertEqual(deployment_2.sub_environments_count, 0)
+
+    def test_deployment_with_invalid_parent_label(self):
+        error_message = 'Deployment invalid_label_dep:' \
+                        ' is referencing deployments using label' \
+                        ' `csys-obj-parent` that are not existed'
```
Comment (geokala) [inline rest-service/manager_rest/test/endpoints/test_deployment_labels_dependencies.py:112]: I recommend not asserting on the entire error message, but instead checking that specific sections are there, e.g. in this case you might check that the error message contained 'invalid_label_dep', 'not', 'exist'.
That allows some rejigging of error messages without triggering test failures, while still complaining if the generally important information isn't included in them.

File: rest-service/manager_rest/resource_manager.py (hunk 1 above)
Comment (mabuaisha) [reply rest-service/manager_rest/resource_manager.py:62]: done

File: rest-service/manager_rest/resource_manager.py (hunk 2 above)
Comment (mabuaisha) [reply rest-service/manager_rest/resource_manager.py:1411]: done

File: rest-service/manager_rest/resource_manager.py (hunk 3 above)
Comment (mabuaisha) [reply rest-service/manager_rest/resource_manager.py:1412]: done

File: rest-service/manager_rest/resource_manager.py (hunk 4 above)
Comment (mabuaisha) [reply rest-service/manager_rest/resource_manager.py:1417]: @geokala The reason why I'm adding this because the time is not inserted by default unfortunately if you check other models you will notice that and if I omitted this the `created_at` is not null by default https://github.com/cloudify-cosmo/cloudify-manager/blob/master/rest-service/manager_rest/storage/resource_models.py#L73 and I do not think so it set by default

File: rest-service/manager_rest/rest/resources_v3_1/deployments.py (hunk 5 above)
Comment (mabuaisha) [reply rest-service/manager_rest/rest/resources_v3_1/deployments.py:114]: Bu mistake I think while was doing rebase and merge conflict but this should be fixed

File: rest-service/manager_rest/rest/resources_v3_1/deployments.py (hunk 6 above)
Comment (mabuaisha) [reply rest-service/manager_rest/rest/resources_v3_1/deployments.py:133]: Well, just for formatting as all non resource methods aligned together

File: rest-service/manager_rest/rest/rest_utils.py (hunk 7 above)
Comment (mabuaisha) [reply rest-service/manager_rest/rest/rest_utils.py:653]: @geokala actually there is a reason for that which is to lock that deployment https://github.com/cloudify-cosmo/cloudify-manager/blob/RD-1485-update-deployment-apis/rest-service/manager_rest/rest/rest_utils.py#L669 here thats why `id` is passed instead of the object because I want to update the deployment counts and use `locking`

File: rest-service/manager_rest/rest/rest_utils.py (hunk 8 above)
Comment (mabuaisha) [reply rest-service/manager_rest/rest/rest_utils.py:658]: @geokala When I checked the default values for integer counter fields in all our models I see that default value is not zero so I sticked with the same approach and the reason why I'm putting this `or` is to help in summation below and prevent adding NoneType to integer. But I think I;m going to change the default value for all counters for deployment to fallback to zero instead which will make this code more cleaner

File: rest-service/manager_rest/rest/rest_utils.py (hunk 9 above)
Comment (mabuaisha) [reply rest-service/manager_rest/rest/rest_utils.py:690]: Answered above, the same reason for locking deployment

File: rest-service/manager_rest/rest/rest_utils.py (hunk 10 above)
Comment (mabuaisha) [reply rest-service/manager_rest/rest/rest_utils.py:701]: Yeah sure this is something I'm going to update now

File: rest-service/manager_rest/rest/rest_utils.py (hunk 11 above)
Comment (mabuaisha) [reply rest-service/manager_rest/rest/rest_utils.py:697]: sure will fix that

File: rest-service/manager_rest/rest/rest_utils.py (hunk 12 above)
Comment (mabuaisha) [reply rest-service/manager_rest/rest/rest_utils.py:714]: Yup will make default to zero, already left comment above

File: rest-service/manager_rest/rest/rest_utils.py (hunk 13 above)
Comment (mabuaisha) [reply rest-service/manager_rest/rest/rest_utils.py:717]: Sure

File: rest-service/manager_rest/resource_manager.py (hunk 4 above)
Comment (tehasdf) [reply rest-service/manager_rest/resource_manager.py:1417]: if you follow your link, you'll see that it already has a default :P

File: rest-service/manager_rest/storage/resource_models.py (hunk 14 above)
Comment (mabuaisha) [reply rest-service/manager_rest/storage/resource_models.py:648]: Well, I added these two properties here because frontend needs to have the latest execution operation for this deployment which will make them to have these values ready to use instead of query each deployment again to get these operations because currently we exposed them,  So I think it would be better to change their names `latext_execution_ total_operations` for example so that we can tell these are for latest execution ?
@geokala

File: rest-service/manager_rest/test/endpoints/test_deployment_labels_dependencies.py (hunk 15 above)
Comment (mabuaisha) [reply rest-service/manager_rest/test/endpoints/test_deployment_labels_dependencies.py:112]: Sure will update that

File: rest-service/manager_rest/resource_manager.py (hunk 4 above)
Comment (mabuaisha) [reply rest-service/manager_rest/resource_manager.py:1417]: Ohh thats so impressing, @tehasdf this is because I'm using pycharm and it redirect me to old env that has a code with default value

File: resources/rest-service/cloudify/migrations/versions/396303c07e35_5_2_to_5_3.py [hunk 16]
```diff
@@ -80,7 +80,8 @@ def _add_deployment_sub_statuses_and_counters():
         sa.Column(
             'sub_environments_count',
             sa.Integer(),
-            nullable=True
+            nullable=True,
```
Comment (geokala) [inline resources/rest-service/cloudify/migrations/versions/396303c07e35_5_2_to_5_3.py:83]: Should it be nullable? With a default of 0 I think it can be made not nullable?

Comment (mabuaisha) [reply resources/rest-service/cloudify/migrations/versions/396303c07e35_5_2_to_5_3.py:83]: I do not think so it should nullable if its default value is set to 0

File: rest-service/manager_rest/rest/resources_v3_1/deployments.py (hunk 6 above)
Comment (geokala) [reply rest-service/manager_rest/rest/resources_v3_1/deployments.py:133]: Fair enough

File: rest-service/manager_rest/rest/rest_utils.py (hunk 7 above)
Comment (geokala) [reply rest-service/manager_rest/rest/rest_utils.py:653]: OK, makes sense

File: rest-service/manager_rest/storage/resource_models.py (hunk 14 above)
Comment (geokala) [reply rest-service/manager_rest/storage/resource_models.py]: Yeah, I think naming them more accurately would probably be good- it's a bit more wordy, but it also says what it actually means.

File: rest-service/manager_rest/test/endpoints/test_deployment_labels_dependencies.py [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+        )
+        deployment_1 = self.client.deployments.get('parent_1')
+        deployment_2 = self.client.deployments.get('parent_2')
+        self.assertEqual(deployment_1.sub_services_count, 1)
+        self.assertEqual(deployment_1.sub_environments_count, 0)
+        self.assertEqual(deployment_2.sub_services_count, 1)
+        self.assertEqual(deployment_2.sub_environments_count, 0)
+
+    def test_deployment_with_invalid_parent_label(self):
+        error_message = 'label `csys-obj-parent` that do not exist'
```
Comment (geokala) [inline rest-service/manager_rest/test/endpoints/test_deployment_labels_dependencies.py:110]: This error message should say something like "does not exist" rather than "do not exist"- did I miss it in this PR or is it elsewhere?

File: rest-service/manager_rest/test/endpoints/test_deployment_labels_dependencies.py [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+                    {
+                        'csys-obj-parent': 'notexist'
+                    }
+                ],
+                resource_id='invalid_label_dep'
+            )
+
+    def test_deployment_with_valid_and_invalid_parent_labels(self):
+        self.put_deployment(deployment_id='parent_1')
+        error_message = 'label `csys-obj-parent` that do not exist'
```
Comment (geokala) [inline rest-service/manager_rest/test/endpoints/test_deployment_labels_dependencies.py:123]: "does not exist" again

File: rest-service/manager_rest/test/endpoints/test_deployment_labels_dependencies.py (hunk 17 above)
Comment (mabuaisha) [reply rest-service/manager_rest/test/endpoints/test_deployment_labels_dependencies.py:110]: I think you miss it but will update it no worries

File: rest-service/manager_rest/test/endpoints/test_deployment_labels_dependencies.py (hunk 18 above)
Comment (mabuaisha) [reply rest-service/manager_rest/test/endpoints/test_deployment_labels_dependencies.py]: fixed

File: rest-service/manager_rest/storage/resource_models.py (hunk 14 above)
Comment (geokala) [reply rest-service/manager_rest/storage/resource_models.py]: You appear to have removed the default when making them not nullable- I think that'll cause issues with restore?

Comment (mabuaisha) [reply rest-service/manager_rest/storage/resource_models.py]: Sure this is back now and part of the migration file

File: rest-service/manager_rest/resource_manager.py [hunk 19]
```diff
@@ -1356,6 +1358,103 @@ def _finalize_create_deployment(self, deployment: models.Deployment):
...
+        if not parents:
+            return
+        missing_parents = []
+        result = self.sm.list(
+            models.Deployment,
+            include=['id'],
+            filters={'id': lambda col: col.in_(parents)}
+        ).items
+        _existing_parents = [_parent[0] for _parent in result]
+        for parent in parents:
```
Comment (geokala) [inline rest-service/manager_rest/resource_manager.py:1379]: This looks like an ideal candidate for using sets?
missing_parents = set(parents) - set(_existing_parents)

Comment (geokala) [reply rest-service/manager_rest/resource_manager.py:1379]: (I think that's the right way to do it anyway- you have a test for this?)

Comment (mabuaisha) [reply rest-service/manager_rest/resource_manager.py:1379]: Yes there is a test case of missing parents

**Review by geokala — APPROVED** [review]
